src/samVisualize/graphGeneration/cfnTemplateGraphGenerator.ts

```diff
@@ -0,0 +1,150 @@
+                }
+                break
+
+            //  Once the YAML is parsed, a "GetAtt" key will point to an array with two elements.
+            //  We only want the logicalNameOfResource, which lies in the first element of the tuple.
+            //  https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-getatt.html
+            case LinkTypes.GetAtt:
+                graph.createLink(parentNodeName, value[0], LinkTypes.GetAtt)
+                break
+
+            //  Adding a single link, from a "Ref" key
+            //  https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-ref.html
+            case LinkTypes.Ref:
+                graph.createLink(parentNodeName, value, LinkTypes.Ref)
+                break
+
+            //  Extracting an link out of a substitution
+            //  A sub link can point to an array, or a single string.
+            //  We can immediately extract the substituion if it points to a string, otherwise we must continue traversing
+            //  https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-sub.html
+            case LinkTypes.Sub:
+                if (_.isString(value)) {
+                    const substitutions = extractSubstitution(value)
+                    for (const destNodeName of substitutions) {
+                        graph.createLink(parentNodeName, destNodeName, LinkTypes.Sub)
+                    }
+                } else {
+                    // Key must point to an array
+                    for (const element of value) {
+                        traverse(graph, element, parentNodeName)
+                    }
+                }
+                break
+            //  If a key is not a possible link, we continue to traverse for nested links
+            default:
+                // If the value is an array, we traverse each traversable element
+                if (Array.isArray(value)) {
+                    for (const element of value) {
+                        traverse(graph, element, parentNodeName)
+                    }
+                } else {
+                    traverse(graph, value, parentNodeName)
+                }
+        }
+    }
+}
+/**
+ * Generates a graph of the resources contained in a CFN YAML template, provided the template adheres to the `Template Anatomy`
+ * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-anatomy.html
+ * @param inputYaml A string representing a YAML template.
+ * @returns A GraphObject representing the graph, or `undefined` if the input does adhere to the `Template Anatomy`
+ */
+export function generateGraphFromYaml(inputYaml: string): GraphObject | undefined {
+    const templateData = yamlStringToObject(inputYaml)
+    // A graph cannot be generated if the template data is not defined, or not an object
+    if (templateData !== undefined && _.isObjectLike(templateData)) {
```

Comment:
Reading the docs I assume this is identical?

```suggestion
    if (_.isObjectLike(templateData)) {
```

Reply:
Oh, yeah woops. I think it's because I added the object condition after and forgot the undefined check was in lodash `isObjectLike`

Reply:
Hm, I swear I remember thinking that `isObjectLike` checks for `undefined` but I don't think it does. It certainly checks for `null`, but this is not enough.
![image](https://user-images.githubusercontent.com/45087499/126096979-8be10ade-e5fd-42fb-a42f-6cd0792a4eac.png)
![image](https://user-images.githubusercontent.com/45087499/126096999-d9ee9b58-b92e-40d1-9f46-62c185e32968.png)